src/platform/Linux/bluez/Helper.cpp

```diff
@@ -401,7 +400,17 @@ static gboolean BluezCharacteristicWriteFD(GIOChannel * aChannel, GIOCondition a
     isSuccess = true;
 
 exit:
-    return isSuccess ? TRUE : FALSE;
+    if (!isSuccess)
+    {
+        // Returning FALSE from the source callback removes the source object
+        // from the context. In order not to trigger GLib critical warning when
+        // manually removing the source in BluezOTConnectionDestroy() mark it
+        // as no longer active.
+        conn->mC1Channel.mWatch = 0;
```

Comment:
Is there some documentation of what mC1Channel/mWatch is? 
Why is setting this mWatch value to 0 a valid way of 'mark it no longer active'?

Reply:
I see mC1Channel.mWatch being a `conn->mC1Channel.mWatch    = g_io_add_watch(channel,....)`

I am not familiar with watches and documentation seems kind of sparse, but it looks like the logic is described as you add a watch which increases a channel reference count and then you can use g_io_channel_unref to decrease it (is this true?). If so, just setting mWatch to 0 seems to leak things right? or is this an initialization error?

I think this probably needs a bit more explaining. Maybe we just need initializers in struct IOChannel ... mpChannel should start as nullptr and mWatch should start with 0.

Reply:
> I am not familiar with watches and documentation seems kind of sparse

Yes, documentation lacks a lot of things regarding proper handing of watches and callback functions. But it works that way that the `g_io_add_watch()` creates source watch and registers it within the main context, then it returns unique ID greater than zero (see documentation for `g_source_attach`). And here we've got a hard-to-maintain behavior when the callback function returns FALSE. In that case source removes itself from the context (see doc for `GSourceFunc`), so it's not valid any more to call `g_source_remove()`. 

> add a watch which increases a channel reference count and then you can use g_io_channel_unref to decrease it (is this true?)

The issue is not with `g_io_create_watch()` (this one increments ref count of channel) but with `g_source_attach()` (this one returns ID) and `g_source_set_callback()`.

Maybe it will be easier to maintain this whole thing if we will use `g_io_create_watch()` and `g_source_attach()` directly (without using `g_io_add_watch()` wrapper). I will have to check that, but as far as I remember there is always the same issue with callback unregistering itself. Here the situation is easier because the `g_source_remove` is called from main context as well, but in other circumstances we would need mutex to guard `mC1Channel.mWatch`.

Reply:
I think that setting source ID tag to zero inside the callback is correct. Event the GLib documentation has got an example exactly like that: https://www.freedesktop.org/software/gstreamer-sdk/data/docs/latest/glib/glib-The-Main-Event-Loop.html#g-source-is-destroyed The only difference is that here we are calling callback and g_source_remove within the same GLib thread - so there is no need for mutex.

Reply:
The example you posted seems to be an example for 'this will fail, do this instead'.

Reply:
Would a:

```
if (g_source_is_destroyed(g_main_current_source())) {
  return;
}
```

as in the example fix things or does that not work?

Reply:
Maybe you are right that such approach will be better (to add `g_source_is_destroyed(g_main_current_source()` to the callback as well). However, this example is for multi-threaded GTK case. In such case it might happen that someone will destry the widget (`SomeWidget *self`) by calling `some_widget_finalize()` on one thread and in the same time the callback will be called on some other thread - "This will fail in a multi-threaded application if the **widget is destroyed** before the idle handler fires". But this PR is for some other thing - when the `g_source_remove()` is called on already destroyed source. In that case we still need to know whether the source tag ID is valid or not. I think that this example is somehow misleading, because in the "fixed" version the callback always returns `FALSE` (which equals to `G_SOURCE_REMOVE`), however, in the callback there is no `self->idle_id = 0;`, which IMHO should still be there.